providers/pipedrive/read_test.go

```diff
@@ -0,0 +1,210 @@
+package pipedrive
+
+import (
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/amp-labs/connectors"
+	"github.com/amp-labs/connectors/common"
+	"github.com/amp-labs/connectors/test/utils/mockutils"
+	"github.com/amp-labs/connectors/test/utils/mockutils/mockserver"
+	"github.com/amp-labs/connectors/test/utils/testroutines"
+	"github.com/amp-labs/connectors/test/utils/testutils"
+)
+
+func TestRead(t *testing.T) { //nolint:funlen,gocognit,cyclop,maintidx
+	t.Parallel()
+
+	nextPageTest := testutils.DataFromFile(t, "activities.json")
+	leads := testutils.DataFromFile(t, "leads.json")
+
+	ErrResponseBody := `{
+		"success":false,
+		"error":"Scope and URL mismatch",
+		"errorCode":403,
+		"error_info":"Please check developers.pipedrive.com"
+	}`
+
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("Content-Type", "application/json")
+		w.WriteHeader(http.StatusOK)
+		_, _ = w.Write(nextPageTest)
+	}))
+
+	tests := []testroutines.Read{
+		{
+			Name:         "Object Name Required",
+			Server:       mockserver.Dummy(),
+			ExpectedErrs: []error{common.ErrMissingObjects},
+		},
+		{
+			Name:         "At Least One Read Field Required",
+			Input:        common.ReadParams{ObjectName: "activities"},
+			Server:       mockserver.Dummy(),
+			ExpectedErrs: []error{common.ErrMissingFields},
+		},
+		{
+			Name:         "NextPage URL construction",
+			Input:        common.ReadParams{ObjectName: "activities", Fields: connectors.Fields("id")},
+			Server:       server,
+			ExpectedErrs: nil,
+			Expected: &common.ReadResult{
+				Rows: 1,
+				Data: []common.ReadResultRow{
+					{
+						Fields: map[string]any{
+							"id": float64(2),
+						},
+						Raw: map[string]any{
+							"id":                            float64(2),
+							"company_id":                    float64(13313052),
+							"user_id":                       float64(20580207),
+							"done":                          false,
+							"type":                          "call",
+							"reference_type":                nil,
+							"reference_id":                  nil,
+							"conference_meeting_client":     nil,
+							"conference_meeting_url":        nil,
+							"due_date":                      "2024-10-30",
+							"due_time":                      "",
+							"duration":                      "",
+							"busy_flag":                     false,
+							"add_time":                      "2024-10-16 12:16:02",
+							"marked_as_done_time":           "",
+							"last_notification_time":        nil,
+							"last_notification_user_id":     nil,
+							"notification_language_id":      nil,
+							"subject":                       "I usually can't come up with words",
```

Comment:
Same :smile: :wink: 

Speaking of `nil` values. You can remove all of them and modify the equality operation to check that all the fields you mention must be present in response. Just like you do in the test below.
```
			Comparator: func(baseURL string, actual, expected *common.ReadResult) bool {
				return mockutils.ReadResultComparator.SubsetFields(actual, expected) &&
					mockutils.ReadResultComparator.SubsetRaw(actual, expected) &&
					actual.NextPage.String() == expected.NextPage.String() &&
					actual.Rows == expected.Rows &&
					actual.Done == expected.Done
			}
```
I will be updating some utilities later on to have this as single utility function replacing several such usages.